### PR TITLE
feat: add builtin tiktoken tokenizer support

### PR DESCRIPTION
**Pull request by matthewkotila**
# Builtin tiktoken tokenizer support

## Origin

Bulk of the implementation cherry-picked from [`ajc/tokenizer-cache`](https://github.com/ai-dynamo/aiperf/commits/ajc/tokenizer-cache/) (commit [`346fbf3f`](https://github.com/ai-dynamo/aiperf/commit/346fbf3f8e343446cd543bc72ee80126d525ec9a) by @ajcasagrande).

## Summary

Add `--tokenizer builtin` option backed by tiktoken's `o200k_base` encoding for zero-network-access token counting. Also adds automatic HF cache detection to skip unnecessary network calls. Includes code review fixes and comprehensive test coverage for the new functionality.

### Commit 1: Cherry-pick from ajc/tokenizer-cache (@ajcasagrande)

Core implementation:

- [x] Add `tiktoken>=0.7.0,<1` dependency to `pyproject.toml`
- [x] `_TiktokenAdapter` class bridging tiktoken's API to `Tokenizer` interface (`tokenizer.py`)
- [x] `_is_hf_cached()` to detect locally cached HF tokenizers and skip network calls (`tokenizer.py`)
- [x] `_find_cached_model_for_alias()` to scan HF cache for alias-style short names (`tokenizer.py`)
- [x] `_from_tiktoken()` factory method on `Tokenizer` with `"builtin"` and encoding name routing (`tokenizer.py`)
- [x] `BUILTIN_TOKENIZER_NAME` and `TIKTOKEN_ENCODING_NAMES` constants for routing decisions (`tokenizer.py`)
- [x] Short-circuit HF alias resolution for tiktoken-backed tokenizers (`tokenizer_validator.py`)
- [x] Update `--tokenizer` CLI description to mention `builtin` option (`tokenizer_config.py`)
- [x] Pre-cache `o200k_base` encoding in Docker image, set `TIKTOKEN_CACHE_DIR` (`Dockerfile`)
- [x] MIT license attribution for OpenAI tiktoken encoding data (`ATTRIBUTIONS.md`)
- [x] Docs for builtin tokenizer, cache detection, updated alias resolution numbering (`tokenizer-auto-detection.md`)
- [x] `TestBuiltinTokenizer` tests and `_no_cache_shortcut` fixture (`test_tokenizer.py`, `test_tokenizer_alias_resolution.py`)

### Commit 2: Self-review fixes and additional tests (@matthewkotila)

Code review fixes:

- [x] Rename `_TIKTOKEN_ENCODING_NAMES` -> `TIKTOKEN_ENCODING_NAMES` (was private but imported cross-module in `tokenizer_validator.py`)
- [x] Fix duplicate "4." numbering in alias resolution docs (`tokenizer-auto-detection.md`)

New test coverage:

- [x] `test_tokenizer_cache.py` (new, 10 tests)
  - `TestIsHfCached`: cache dir missing, exact match, alias match (case-insensitive), no match, non-model dirs ignored, empty cache
  - `TestFindCachedModelForAlias`: finds alias, no match, cache missing, case-insensitive
- [x] `test_tokenizer.py` (extended, 9 new tests)
  - `TestTiktokenEncodingNames`: parametrized over all 6 encoding names + direct `o200k_base` equivalence with `builtin`
  - `TestTiktokenImportError`: verifies `TokenizerError` when tiktoken isn't installed
- [x] `test_tokenizer_validator.py` (new, 7 tests)
  - `TestValidatorTiktokenShortCircuit`: `"builtin"` + parametrized over all 6 encoding names, verifying `resolve_alias` is never called

## Files changed

| File | Change |
|---|---|
| `pyproject.toml` | Add `tiktoken>=0.7.0,<1` dependency |
| `src/aiperf/common/tokenizer.py` | `_TiktokenAdapter`, `_is_hf_cached()`, `_find_cached_model_for_alias()`, builtin/tiktoken routing, `TIKTOKEN_ENCODING_NAMES` |
| `src/aiperf/common/tokenizer_validator.py` | Short-circuit HF resolution for tiktoken tokenizers |
| `src/aiperf/common/config/tokenizer_config.py` | Update `--tokenizer` description to mention `builtin` |
| `Dockerfile` | Pre-cache `o200k_base` encoding, set `TIKTOKEN_CACHE_DIR` |
| `ATTRIBUTIONS.md` | MIT license attribution for OpenAI tiktoken encoding data |
| `docs/cli-options.md` | Auto-generated CLI docs update |
| `docs/reference/tokenizer-auto-detection.md` | Builtin tokenizer docs, cache detection docs, fix numbering |
| `tests/unit/common/test_tokenizer.py` | `TestBuiltinTokenizer`, `TestTiktokenEncodingNames`, `TestTiktokenImportError` |
| `tests/unit/common/test_tokenizer_alias_resolution.py` | `_no_cache_shortcut` autouse fixture |
| `tests/unit/common/test_tokenizer_cache.py` | `TestIsHfCached`, `TestFindCachedModelForAlias` |
| `tests/unit/common/test_tokenizer_validator.py` | `TestValidatorTiktokenShortCircuit` |

## Verification

- [x] Pre-commit hooks pass (`pre-commit run --all-files`)
- [x] All 133 tokenizer-related unit tests pass
- [x] Full unit test suite passes (`uv run pytest tests/unit/ -n auto`)

## Summary by CodeRabbit

* **New Features**
  * Added `--tokenizer builtin` for a zero-network-access tokenizer (o200k_base) and support for loading tokenizers from a local cache.

* **Documentation**
  * Updated CLI and tokenizer reference docs to describe builtin mode and automatic local-cache detection.
  * Added attribution noting included MIT-licensed tokenizer data.

* **Tests**
  * Added tests covering builtin tokenizer behavior, cache detection, and validation logic.

* **Chores**
  * Added tiktoken as a project dependency and pre-cache support for images.

### CODE REVIEW

File: docs/reference/tokenizer-auto-detection.md
Comment (matthewkotila) [inline docs/reference/tokenizer-auto-detection.md:1]: @coderabbitai (responding to nitpicks here)

tests/unit/common/test_tokenizer_alias_resolution.py:
tests/unit/common/test_tokenizer_cache.py:
tests/unit/common/test_tokenizer_validator.py:

Added return type annotations to all new fixtures.

Dockerfile:

Agreed. Replaced the manual SHA1 derivation with a tiktoken.get_encoding('o200k_base') call during the build, so tiktoken manages its own cache filenames. No longer coupled to its internal caching scheme.

File: ATTRIBUTIONS.md [hunk 1]
```diff
@@ -158,6 +158,40 @@ This document provides attribution information for public domain assets used in
...
+## Tokenizer Data
+
+### OpenAI tiktoken Encoding Data
+
+**Asset Information:**
+- **File(s)**: `o200k_base.tiktoken` (pre-cached in Docker image at `/opt/tiktoken_cache/`)
+- **Source**: [openai/tiktoken](https://github.com/openai/tiktoken) via `https://openaipublic.blob.core.windows.net/encodings/o200k_base.tiktoken`
+- **Author**: OpenAI and Shantanu Jain
+- **License**: MIT License
+
```
[the remark this answers is not in the document]
Comment (matthewkotila) [reply ATTRIBUTIONS.md:170]: Good catch--updated the usage summary to distinguish public domain assets from MIT-licensed ones.

File: docs/reference/tokenizer-auto-detection.md [hunk 2]
```diff
@@ -17,12 +17,27 @@ AIPerf resolves tokenizer names **before** spawning services via lightweight Hub
...
 
+## Built-in Tokenizer
+
+Pass `--tokenizer builtin` to use a zero-network-access tokenizer backed by [tiktoken](https://github.com/openai/tiktoken) with the `o200k_base` encoding (GPT-4o / o1 / o3, 200k vocabulary). This skips all HuggingFace Hub alias resolution and downloads.
+
+Use this when you don't need a model-specific tokenizer and just want token counts for performance metrics. The encoding data is downloaded once on first use and cached locally by tiktoken -- subsequent runs require no network access.
+
+```bash
+aiperf run --tokenizer builtin ...
+```
```
[the remark this answers is not in the document]
Comment (matthewkotila) [reply docs/reference/tokenizer-auto-detection.md:28]: Fixed, thanks. The command is `aiperf profile`.

File: src/aiperf/common/tokenizer.py [hunk 3]
```diff
@@ -16,10 +16,56 @@
...
+        return self._encoding.eot_token
+
+    def encode(self, text: str, **kwargs) -> list[int]:
+        return self._encoding.encode(text, allowed_special="all")
+
+    def decode(self, token_ids: list[int], **kwargs) -> str:
+        return self._encoding.decode(token_ids)
+
+    def __call__(self, text: str, **kwargs) -> dict:
+        return {"input_ids": self.encode(text)}
```
[the remark this answers is not in the document]
Comment (matthewkotila) [reply src/aiperf/common/tokenizer.py:61]: Intentionally not changing this. The adapter is internal and no callers pass these kwargs. Hardcoding `allowed_special="all"` is deliberate--aiperf needs to encode all tokens including special ones for accurate token counting. Forwarding kwargs would let callers accidentally break that behavior.

File: src/aiperf/common/tokenizer.py [hunk 4]
```diff
@@ -66,6 +112,34 @@ def _is_offline_mode() -> bool:
...
+        return True
+
+    # Alias match: "gpt2" -> any "models--*--gpt2"
+    suffix = f"--{name.lower()}"
+    return any(
+        entry.is_dir()
+        and entry.name.startswith("models--")
+        and entry.name.lower().endswith(suffix)
+        for entry in cache_dir.iterdir()
+    )
```
[the remark this answers is not in the document]
Comment (matthewkotila) [reply src/aiperf/common/tokenizer.py]: Good edge case. Updated to return `True` only when exactly one alias match exists, so ambiguous names still go through normal resolution. Added a test for it.

File: Dockerfile [hunk 5]
```diff
@@ -133,6 +133,10 @@ RUN mkdir -p /app /app/artifacts /app/.cache \
     && chown -R 1000:1000 /app \
     && chmod -R 755 /app
 
+# Pre-cache tiktoken o200k_base encoding for --tokenizer builtin (MIT license, see ATTRIBUTIONS.md)
+RUN mkdir -p /opt/tiktoken_cache \
```
Comment (FrankD412) [inline Dockerfile]: @matthewkotila -- question; why does the `import tiktoken` below work? We're installing `aiperf` later, and the build seems to be based on a slim container. I'm confused how `tiktoken` is importable at this point.

Automated review found the same thing:

  Inline comment 1 — Dockerfile:136-142 (CRITICAL)

  **Bug: tiktoken pre-caching runs before tiktoken is installed**

  The `python -c "import tiktoken"` at line 138 will fail with
  `ModuleNotFoundError` because `uv sync` (which installs tiktoken from
  `pyproject.toml`) doesn't run until line 142.

  The `env-builder` stage inherits from `base`, which creates an empty venv —
  no packages are installed until `uv sync`.

  Fix: move the tiktoken pre-caching `RUN` block to after the `uv pip install` step:

  ```dockerfile
  RUN uv pip install /dist/aiperf-*.whl \
      && rm -rf /dist /workspace/pyproject.toml

  # Pre-cache tiktoken o200k_base encoding (MIT license, see ATTRIBUTIONS.md)
  RUN mkdir -p /opt/tiktoken_cache \
      && TIKTOKEN_CACHE_DIR=/opt/tiktoken_cache python -c \
         "import tiktoken; tiktoken.get_encoding('o200k_base')"

  ---

File: src/aiperf/common/tokenizer.py [hunk 6]
```diff
@@ -66,6 +112,36 @@ def _is_offline_mode() -> bool:
     )
 
 
+def _is_hf_cached(name: str) -> bool:
```
Comment (FrankD412) [inline src/aiperf/common/tokenizer.py:140]: Something Claude found during review:

  ### Inline comment 3 — `src/aiperf/common/tokenizer.py:336-361` (low, optional)

  _find_cached_model_for_alias returns the first filesystem match without a
  uniqueness check, whereas _is_hf_cached (which always runs before this)
  enforces len(matches) == 1. The two are safe together (the pre-check
  guarantees uniqueness), but the inconsistency makes _find_cached_model_for_alias
  hard to reason about in isolation. Consider adding a brief comment:

  # Called only after _is_hf_cached() confirmed exactly one alias match,
  # so the first hit here is always the unique one.


You code tests '_find_cached_model_for_alias` with `_is_hf_cached` but I don't see it called (unless it's what `AliasResolutionResult` utilizes).

File: Dockerfile (hunk 5 above)
Comment (matthewkotila) [reply Dockerfile]: Done

File: src/aiperf/common/tokenizer.py (hunk 6 above)
Comment (matthewkotila) [reply src/aiperf/common/tokenizer.py:140]: Good catch on the inconsistency--I extracted `_find_hf_cache_aliases()` as a shared utility so both functions use the same scan logic and enforce single-match semantics consistently. Added an ambiguity test for `_find_cached_model_for_alias()` too.

**Review by FrankD412 — APPROVED** [review]
